src/main/java/com/es/phoneshop/DAO/CustomProductDao.java

```diff
@@ -79,11 +61,11 @@ public List<Product> findProducts() {
     }
 
     @Override
-    public void save(Product product) throws ProductException {
+    public void save(Product product) throws ProductDefinitionException {
         writeLock.lock();
         try {
             if (product == null)
-                throw new ProductException("Product has no data");
+                throw new ProductDefinitionException("Product has no data");
             Optional<Product> oldProduct = getProductById(product.getId());
```

Comment:
can't we create a deadlock here?
instead of getting getProductById() we can still navigate through stream